Lib/idlelib/autocomplete.py

```diff
@@ -90,14 +93,177 @@ def _delayed_open_completions(self, args):
+
+    @classmethod
+    def _is_completing_dict_key(cls, hyper_parser):
+        hp = hyper_parser
+        if hp.is_in_string():
+            # hp.indexbracket is an opening quote;
+            # check if the bracket before it is '['.
+            if hp.indexbracket == 0:
+                return None
+            opening_bracket_idx = hp.bracketing[hp.indexbracket - 1][0]
+            if hp.rawtext[opening_bracket_idx] != '[':
+                return None
+            # The completion start is everything from the '[', except initial
+            # whitespace. Check that it's a valid beginning of a str/bytes
+            # literal.
+            opening_quote_idx = hp.bracketing[hp.indexbracket][0]
+            m = cls._dict_str_key_prefix_re.fullmatch(
+                hp.rawtext[opening_bracket_idx+1:opening_quote_idx])
+            if m is None:
+                return None
+            comp_start = \
+                m.group(1) + hp.rawtext[opening_quote_idx:hp.indexinrawtext]
+        elif hp.is_in_code():
+            # This checks the the last bracket is an opener,
+            # and also avoids an edge-case exception.
+            if not hp.isopener[hp.indexbracket]:
+                return None
+            # Check if the last opening bracket a '['.
+            opening_bracket_idx = hp.bracketing[hp.indexbracket][0]
+            if hp.rawtext[opening_bracket_idx] != '[':
+                return None
+            # We want to complete str/bytes dict keys only if the existing
+            # completion start could be the beginning of a literal. Note
+            # that an empty or all-whitespace completion start is acceptable.
+            m = cls._dict_str_key_prefix_re.fullmatch(
+                hp.rawtext[opening_bracket_idx + 1:hp.indexinrawtext])
+            if m is None:
+                return None
+            comp_start = m.group(1)
+        else:
+            return None
+
+        hp.set_index("insert-%dc" % (len(hp.rawtext) - opening_bracket_idx))
+        comp_what = hp.get_expression()
+        return comp_what, comp_start
+
+    @classmethod
+    def _dict_key_reprs(cls, comp_start, comp_list):
+        """Customized repr() for str and bytes.
+
+        This retains the prefix and quote type of the completed partial
+        str/bytes literal, if they exist.
+        """
+        # Analyze existing prefix and quote type of the completion start.
+        is_raw = False
+        quote_type = '"'
+        prefix_match = cls._dict_str_key_prefix_re.match(comp_start)
+        if prefix_match is not None:
+            prefix = prefix_match.group(1)
+            post_prefix = comp_start[prefix_match.end():]
+
+            is_raw = 'r' in prefix.lower()
+            if post_prefix:
+                quote_type_candidate = post_prefix[0]
+                if quote_type_candidate not in "'\"":
+                    pass  # use the default quote type
+                elif post_prefix[1:2] == quote_type_candidate:
```

Comment:
> When completion is invoked after an opening quote, it is clear that a string completion is desired, and perhaps then only string/bytes completions should be offered.

@taleinat Could this be done by filtering ``comp_list`` after ``quote_type_candidate`` is determined to be either ``'`` or ``"``?
```suggestion
                comp_list = [c for c in comp_list if c not in globals()]
                elif post_prefix[1:2] == quote_type_candidate:
```
Or is there a better location to do this?